source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+// calculates the result for each element of the result vector to return.
+// Vectors with longer than 32-bit scalar components are not accepted in this
+// function.
+std::vector<uint32_t> OperateVectors(
+    SpvOp opcode, uint32_t num_dims,
+    const std::vector<analysis::Const*>& operands) {
+  std::vector<uint32_t> result;
+  for (uint32_t d = 0; d < num_dims; d++) {
+    std::vector<uint32_t> operand_values_for_one_dimension;
+    for (analysis::Const* operand : operands) {
+      if (analysis::VectorConst* vector_operand = operand->AsVectorConst()) {
+        if (const analysis::ScalarConst* scalar_component =
+                vector_operand->GetComponent(d)->AsScalarConst()) {
+          const auto& scalar_words = scalar_component->words();
+          assert(
+              scalar_words.size() == 1 &&
+              "Vector components with longer than 32-bit width are not allowed "
+              "in OperateVectors()");
+          operand_values_for_one_dimension.push_back(scalar_words.front());
+        } else if (operand->AsNullConst()) {
+          operand_values_for_one_dimension.push_back(0u);
+        } else {
+          assert(false &&
+                 "VectorConst should only has ScalarConst or NullConst as "
+                 "components");
+        }
+      } else if (operand->AsNullConst()) {
+        operand_values_for_one_dimension.push_back(0u);
+      } else {
+        assert(false &&
+               "OperateVectors() only accepts VectorConst or NullConst type of "
+               "constant");
+      }
+    }
+    result.push_back(OperateWords(opcode, operand_values_for_one_dimension));
+  }
+  return result;
+}
+}  // anonymous namespace
+
+bool FoldSpecConstantOpAndCompositePass::ProcessImpl(ir::Module* module) {
+  bool modified = false;
+  auto inst_iter = module->types_values_begin();
+  // In case the types, constants, global variabls secion is empty, do nothing.
+  if (inst_iter == module->types_values_end()) return modified;
+
+  do {
+    // Collect constant values of normal constants and process the
+    // OpSpecConstantOp and OpSpecConstantComposite instructions if possible.
+    // OpConstantSampler instruction is not collected here because it cannot
+    // be used in OpSpecConstant{Composite|Op} instructions.
+    ir::Instruction* inst = &*inst_iter;
+    switch (SpvOp opcode = inst->opcode()) {
+      case SpvOp::SpvOpConstantTrue:
+      case SpvOp::SpvOpConstantFalse:
+      case SpvOp::SpvOpConstant:
+      case SpvOp::SpvOpConstantNull:
+      case SpvOp::SpvOpConstantComposite:
+      case SpvOp::SpvOpSpecConstantComposite: {
+        if (auto const_value = CreateConstFromInst(inst)) {
+          // Need to replace the OpSpecConstantComposite instruction with a
+          // corresponding OpConstantComposite instruction.
+          if (opcode == SpvOp::SpvOpSpecConstantComposite) {
+            inst->SetOpcode(SpvOp::SpvOpConstantComposite);
```

Comment:
When do the operands of inst get replaced/rewritten?   Is it handled through the def-use manager?
Should say so.